### ISSUE TITLE
Test runs leave behind files from headless host storage

### ISSUE THREAD

**Issue by bdach** [ppy/osu#8420]
Running a test full test suite locally can result in having lots of leftover files in the execution directory, which can build up to as much as gigabytes - at least on Windows - if left unattended for long enough (personally racked up about ~25 GB).

Some of this is due to the game, some of it due to the framework. For instance, the framework does not dispose of font stores in `Game`, therefore i.e. holding references to the font cache when the headless host is cleaning up, and as such the cleanup partially fails due to the file being open by the same process, leaving behind a footprint of 1 MB per instantiated host. osu!-side, sqlite database files also stay behind after test runs, making up about ~200 KB per each test host.

Those are the two main cases of leftover storage I can see - there might be more.

I attempted a *very* simplistic fix by changing framework's disposal in `Game` to this:

```csharp
        protected override void Dispose(bool isDisposing)
        {
            Audio?.Dispose();
            Audio = null;

            Fonts?.Dispose();
            Fonts = null;

            localFonts?.Dispose();
            localFonts = null;

            Resources?.Dispose();
            Resources = null;

            Textures?.Dispose();
            Textures = null;

            base.Dispose(isDisposing);
        }
```

but it only fixes the framework side part of the problem and additionally causes test failures game-side that I haven't been able to investigate in detail yet, therefore I figured I'd open an issue even if I eventually get to resolving this myself.

The overall fingerprint of leftover headless host storages in one full game-side run is about 1.2 GB.

**Comment by peppy** [comment]
This may be specific to the test runner / platform you're using. On macOS the only files left behind are the occasional logs (~200bytes per run). Are you using `dotnet test` or an nunit IDE plugin to run? Also which OS?

**Comment by bdach** [comment]
I'm seeing this particular behaviour on Windows on unit runs, both by `dotnet test` and through rider.

**Comment by peppy** [comment]
I think fixing the framework side would be a good start.

**Comment by bdach** [comment]
The framework side is fixed now, but the db file thing isn't. Generally I noticed that the recently-introduced [`FlushConnections()`](https://github.com/ppy/osu/blob/b0cd5c1a08b87d7ce5b24260943f7aa47b72a3f9/osu.Game/Database/DatabaseContextFactory.cs#L164-L170) method is way better at freeing up the database files, so I attempted to use it in [`ResetDatabase()`](https://github.com/ppy/osu/blob/b0cd5c1a08b87d7ce5b24260943f7aa47b72a3f9/osu.Game/Database/DatabaseContextFactory.cs#L147-L162) directly, but that didn't fix it completely.

Storages in test scenes derived from `OsuTestScene` are super confusing because as far as I can tell when running via `dotnet test`, they for some reason create *two* storages - one in `headless/{sceneName}-{guid}` (coming from the framework), one in `./{sceneName}-{guid}` (directly from `OsuTestScene`). Some of the test scenes game-side still hold font caches open therefore partially invalidating the framework-side fix. I'm guessing this is due to some borked lifetimes of the font stores, but I don't have a clear explanation what's going on yet.

Leaving this on the backburner for now as the situation should be quite improved already, but it's not totally fixed yet. DB migrations might also render this moot.
